paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java

```diff
@@ -0,0 +1,270 @@
+import org.apache.paimon.table.sink.CommitMessage;
+import org.apache.paimon.table.sink.CommitMessageImpl;
+import org.apache.paimon.utils.SnapshotManager;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.stream.Collectors;
+
+/** {@link AppendOnlyFileStoreTable} compact coordinator. */
+public class AppendOnlyTableCompactionCoordinator {
+    private final SnapshotManager snapshotManager;
+    private final AppendOnlyFileStoreScan scan;
+    private Long snapshotId;
+
+    private final long targetFileSize;
+    private final int minFileNum;
+    private final int maxFileNum;
+
+    private final Map<BinaryRow, PartitionCompactCoordinator> partitionCompactCoordinators =
+            new HashMap<>();
+
+    public AppendOnlyTableCompactionCoordinator(
+            AppendOnlyFileStoreTable table, long targetFileSize, int minFileNum, int maxFileNum) {
+        scan = table.store().newScan();
+        snapshotManager = table.snapshotManager();
+        this.targetFileSize = targetFileSize;
+        this.minFileNum = minFileNum;
+        this.maxFileNum = maxFileNum;
+        snapshotId = snapshotManager.latestSnapshotId();
+
+        if (snapshotId != null) {
+            updateFull();
+        }
+    }
+
+    public void updateRestore() {
+        Long latestSnapshotId = snapshotManager.latestSnapshotId();
+        if (latestSnapshotId == null || latestSnapshotId.equals(snapshotId)) {
+            return;
+        }
+
+        if (snapshotId == null) {
+            snapshotId = latestSnapshotId;
+            updateFull();
+        } else {
+            updateDelta(latestSnapshotId);
+            snapshotId = latestSnapshotId;
+        }
+    }
+
+    private void updateFull() {
+        scan.withSnapshot(snapshotId).plan().files().stream()
+                .collect(Collectors.groupingBy(ManifestEntry::partition))
+                .forEach(
+                        (k, v) ->
+                                partitionCompactCoordinators
+                                        .computeIfAbsent(k, PartitionCompactCoordinator::new)
+                                        .addRestoredFiles(
+                                                v.stream()
+                                                        .map(ManifestEntry::file)
+                                                        .filter(
+                                                                dataFileMeta ->
+                                                                        dataFileMeta.fileSize()
+                                                                                < targetFileSize)
+                                                        .collect(Collectors.toList())));
+    }
+
+    private void updateDelta(Long targetSnapshotId) {
```

Comment:
We don't need to `updateDelta`, just update full and store delta files from writers.